src/internal/storage/chunk/uploader.go

```diff
@@ -76,6 +81,97 @@ func (u *Uploader) Upload(meta interface{}, r io.Reader) error {
 }
 
 func (u *Uploader) Copy(meta interface{}, dataRefs []*DataRef) error {
+	var stableDataRefs, nextDataRefs []*DataRef
+	taskChain := taskchain.New(u.ctx, u.chunkSem)
+	appendDataRefs := func(dataRefs []*DataRef) error {
+		return taskChain.CreateTask(func(_ context.Context) (func() error, error) {
+			return func() error {
+				for _, dataRef := range dataRefs {
+					stableDataRefs = append(stableDataRefs, proto.Clone(dataRef).(*DataRef))
+				}
+				return nil
+			}, nil
+		})
+	}
+	for i := 0; i < len(dataRefs); i++ {
+		if !stableDataRef(dataRefs[i]) {
+			if err := appendDataRefs(nextDataRefs); err != nil {
+				return err
+			}
+			nextDataRefs = nil
+			var err error
+			i, err = u.align(u.ctx, dataRefs, i, func(chunk []byte) error {
+				return taskChain.CreateTask(func(ctx context.Context) (func() error, error) {
+					dataRef, err := upload(ctx, u.client, chunk, nil, u.noUpload)
+					if err != nil {
+						return nil, err
+					}
+					return func() error {
+						stableDataRefs = append(stableDataRefs, dataRef)
+						return nil
+					}, nil
+				})
+			})
+			if err != nil {
+				return err
+			}
+			continue
+		}
+		nextDataRefs = append(nextDataRefs, dataRefs[i])
+	}
+	if err := appendDataRefs(nextDataRefs); err != nil {
+		return err
+	}
+	return u.taskChain.CreateTask(func(_ context.Context) (func() error, error) {
+		if err := taskChain.Wait(); err != nil {
+			return nil, err
+		}
+		if len(dataRefs) > 0 {
+			stableDataRefs[0].Ref.Edge = true
+			stableDataRefs[len(stableDataRefs)-1].Ref.Edge = true
+		}
+		return func() error {
+			return u.cb(meta, stableDataRefs)
+		}, nil
+	})
+}
+
+// A data reference is stable if it does not reference an edge chunk and references the full chunk.
+func stableDataRef(dataRef *DataRef) bool {
+	return !dataRef.Ref.Edge && dataRef.OffsetBytes == 0 && dataRef.SizeBytes == dataRef.Ref.SizeBytes
+}
+
+func (u *Uploader) align(ctx context.Context, dataRefs []*DataRef, i int, cb func([]byte) error) (int, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	if err := miscutil.WithPipe(func(w2 io.Writer) error {
+		r := u.storage.NewReader(ctx, dataRefs[i:], WithPrefetchLimit(3))
+		return r.Get(w2)
+	}, func(r io.Reader) error {
+		splitBytesLeft := dataRefs[i].SizeBytes
+		return ComputeChunks(r, func(chunk []byte) error {
+			chunkBytesLeft := int64(len(chunk))
+			if err := cb(chunk); err != nil {
+				return err
+			}
+			for chunkBytesLeft > splitBytesLeft {
+				chunkBytesLeft -= splitBytesLeft
+				i++
+				splitBytesLeft = dataRefs[i].SizeBytes
+			}
+			if chunkBytesLeft == splitBytesLeft {
+				return errutil.ErrBreak
+			}
+			splitBytesLeft -= chunkBytesLeft
+			return nil
+		})
+	}); err != nil && !errors.Is(err, errutil.ErrBreak) {
+		return 0, err
+	}
+	return i, nil
+}
+
+func (u *Uploader) CopyByReference(meta interface{}, dataRefs []*DataRef) error {
```

Comment:
Should this be public?  If it's possible to misuse this, then it should be private and we can call into it in Copy

Reply:
This comment might be misplaced:  Somewhere in here there needs to be a call to client in order to create the dependency between the chunks.  That is also the mechanism that prevents dangling references.

Reply:
We discussed this offline. This can only be misused in the context of the client (i.e. incorrect file hashes). A client can choose to either perform a normal copy when they want to ensure stable data references or use copy by reference when they just want to reuse the references as is.